lib/fhir_models/validation/element_validator/code_validator.rb

```diff
@@ -0,0 +1,115 @@
+module FHIR
+  module Validation
+    # Validates a coded element
+    class CodeValidator
+      attr_accessor :results
+      attr_reader :element_definition, :element, :path, :validator
+
+      def initialize(element, element_definition, path, validator)
+        @element = element
+        @element_definition = element_definition
+        @path = path
+        @validator = validator
+      end
+
+      # Validate a coded element
+      #
+      # @return result [Array<FHIR::ValidationResult>] the result of the data type validation
+      def validate
+        results = []
+        if %w[CodeableConcept Coding Quantity].include? type_code
+          if type_code == 'CodeableConcept'
+            element.coding.each do |coding|
+              results.concat(check_code(coding, valueset_uri))
+            end
+          else
+            # avoid checking Codings twice if they are already checked as part of a CodeableConcept
+            # The CodeableConcept should contain the binding for the children Codings
+            results.concat(check_code(element, valueset_uri)) unless element_definition.path.include? 'CodeableConcept.coding'
+          end
+        end
+        results
+      end
+
+      # Returns the type of the element based off the element definition and path
+      #
+      # This is useful for determining the type when a choice of types are present.
+      #
+      # @return type_code [String] the type of the element
+      private def type_code
+        element_definition.type_code(path)
+      end
+
+      # If the binding is set to 'required'
+      #
+      # @return required_binding? [Boolean] if the binding is set to 'required'
+      private def required_binding?
+        element_definition&.binding&.strength == 'required'
+      end
+
+      # The URI of the bound ValueSet
+      #
+      # @return valueset_uri [String]
+      private def valueset_uri
+        element_definition&.binding&.valueSet
+      end
+
+      # if failures return a warn of fail result
+      #
+      # @return fail_level [:fail, :warn]
+      private def fail_level
+        required_binding? ? :fail : :warn
+      end
+
+      # Check that both a code and system are present in the coded element
+      # @param coding [#code & #system]
+      # @return [Array<FHIR::ValidationResult>]
+      private def check_for_code_and_system(coding)
+        check_results = []
+        check_results.push(new_result(:warn, "#{path}: missing code")) if coding.code.nil?
```

Comment:
If the binding is `required`, but no `code` is present I don't think that would be a `:fail`.

[`required`](https://www.hl7.org/fhir/terminologies.html#strength) binding indicates that:

 `To be conformant, the concept in this element SHALL be from the specified value set.`

If there is no concept in the element then we shouldn't fail even if the binding is required (a cardinality failure is a different matter).  I talked to @Jammjammjamm and it might make sense to not return any result here as there is nothing to validate if there is no code.

Reply:
Yes, agreed. If the code and system are both empty, there is nothing to say.

If they are required, they'll be caught by the cardinality checks.